src/ophyd_async/core/soft_signal_backend.py

```diff
@@ -71,7 +75,20 @@ def make_initial_value(self, datatype: Optional[Type[T]]) -> T:
 
 class SoftArrayConverter(SoftConverter):
     def get_datakey(self, source: str, value, **metadata) -> DataKey:
-        return {"source": source, "dtype": "array", "shape": [len(value)], **metadata}
+        dtype_numpy = ""
+        if isinstance(value, list):
+            if len(value) > 0:
+                dtype_numpy = np.dtype(type(value)).descr[0][1]
```

Comment:
I'm not sure I understand how this works. Should it be `np.dtype(value[0].dtype)` on line 81 (or something along those lines). Also, is value always either a scalar or a _list_? Would it be safer to check if it's an iterable instead?
Or is it a numpy array or a list?
Maybe adding a typehint to `value` in the function signature could be helpful.

Reply:
Good points, @genematx! I overlooked that case during my review.
Maybe it should just be `np.dtype(type(value[0]))` as the `value[0]` will be a number, which won't (or may not?) have `.dtype` method.

To compare two approaches:
```py
In [28]: np.dtype(type([1, 2, 3][0])).descr
Out[28]: [('', '<i8')]

In [29]: np.dtype(type([1, 2, 3])).descr
Out[29]: [('', '|O')]
```

I support adding a type annotation to the signature and a docstring to explain the expected input types.